File: LeetCode/211DesignAddAndSearchWordsDataStructure/DesignAddAndSearchWordsDataStructure.cpp
struct TrieNode {
    map<char,TrieNode*> children;
    bool endOfWord = false;
};


class WordDictionary {
public:
    TrieNode* root = new TrieNode();
    WordDictionary() {
        
    }
    
    void addWord(string word) {
        TrieNode *current = root;
        for (int i = 0; i < word.size(); i++) {
            if (current->children.find(word[i]) == current->children.end()) {
                current->children[word[i]] = new TrieNode();
            }
            current = current->children[word[i]];
        }
        current->endOfWord = true;
    }

    bool dfs(string &word, int startIndex, TrieNode* node) {
        TrieNode *current = node;
        for (int i = startIndex; i < word.size(); i++) {
            if (word[i] == '.') {
                for (pair<const char, TrieNode *> key:current->children) {
                    if (dfs(word, i + 1, current->children[key.first])) {
                        return true;
                    }
                }
                return false;
            }
            else {
                if (current->children.find(word[i]) == current->children.end()) {
                    return false;
                }
                current = current->children[word[i]];
            }
        }
        return current->endOfWord;
    }
    
    bool search(string word) {
        return dfs(word, 0, root);
    }
};

File: LeetCode/211DesignAddAndSearchWordsDataStructure/DesignAddAndSearchWordsDataStructure.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.endOfWord = False

class WordDictionary:

    def __init__(self):
        self.root = TrieNode()

    def addWord(self, word: str) -> None:
        current = self.root
        for letter in word:
            if letter not in current.children:
                current.children[letter] = TrieNode()
            current = current.children[letter]
        current.endOfWord = True

    def search(self, word: str) -> bool:
        def dfs(startIndex, root):
            current = root
            for i in range(startIndex, len(word)):
                letter = word[i]
                if letter == ".":
                    for child in current.children.values():
                        if dfs(i+1, child):
                            return True
                    return False
                else:
                    if letter not in current.children:
                        return False
                    current = current.children[letter]
            return current.endOfWord
        return dfs(0, self.root)
